[결제] 결제하기 기능 추가 이벤트 기반 비즈니스 로직만 작성

zmfpdl64 opened the pull request:
## 개요
이벤트 기반으로 구현하기 위해 결제했을 때의 로직과 이벤트중 실패했을 때의 롤백 로직을 구현했습니다.
테스트 코드는 다른 PR로 올리겠습니다

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [x] 주석 추가 및 수정
- [x] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정
- [x] 파일 혹은 폴더명 수정

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  Commit message convention 참고  (Ctrl + 클릭하세요.) 
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

> ### Description

- [x] 수정해야할 코드 Refactor ✅ 2024-05-08
- [x] 결제 기능 ✅ 2024-05-08
    - [x] 서비스 ✅ 2024-05-08
        - [x] 결제할 때 ✅ 2024-05-08
        - [x] 롤백할 때 ✅ 2024-05-08
    - [x] 예외 ✅ 2024-05-08
    - [x] Dto ✅ 2024-05-08
- [x] Internal external Controller 분리 ✅ 2024-05-05
- [x] 멀티분산락  ✅ 2024-05-05

## 결제 로직
1. 유저 빈페이 가져온다
2. 판매자 빈페이들을 가져온다.
3. 판매자 빈페이와 request를 맵핑한다
4. 결제를 진행한다
	1. 결제 디테일을 생성한다
		1. 디테일 상태 히스토리를 생성한다
	2. 유저가 돈을 지불한다
		1. 빈페이 디테일이 생성된다
		2. 유저와, 셀러의 빈페이 값이 변경된다


## 결제 롤백 로직

1. 결제를 가져온다.
2. 결제객체의 롤백 로직을 호출한다.
	1. 결제 디테일과 빈페이 디테일 상태값을 변경한다
	    1. 상세 히스토리를 생성한다
	2. 유저간의 돈을 롤백한다.

> ### Core Code

```java
	@DistributedLock(
		// name = "BEANPAY",
		key = {
			"#paymentPrice.userId().concat('USER')",
			"#paymentPrice.paymentDetails()"
		}
	)
	public PaymentDto paymentPrice(final PaymentPrice paymentPrice) {
		//유저 BeanPay 가져오기
		final BeanPay userBeanPay = getBeanPay(paymentPrice.userId(), USER);

		// 판매자들 BeanPay 가져오기
		final List<BeanPay> sellerBeanPays =
			beanPayRepository.findBeanPayByUserIdsAndRole(
			paymentPrice.getSellerIds(),
			SELLER
		);

		// 결제
		final Payment payment = Payment.ofPayment(
			userBeanPay,
			paymentPrice.orderId(),
			paymentPrice.totalAmount(),
			paymentPrice.orderName(),
			mappedBeanPayPaymentDetailPrice(paymentPrice, sellerBeanPays)
		);

		Payment save = paymentRepository.save(payment);
		return PaymentMapper.INSTANCE.toDto(save);
	}
```
싱글락에서 멀티락도 호환할 수 있도록 구현했습니다

zmfpdl64 committed:
Refactor: external internal 분리

zmfpdl64 committed:
Feat: 멀티분산락 구현

zmfpdl64 committed:
Refactor: type 폴더 삭제

zmfpdl64 committed:
Test: 멀티분산락 테스트 작성

zmfpdl64 committed:
Feat: 결제 DTO 구현

zmfpdl64 committed:
Feat: Enum 상태값 추가

zmfpdl64 committed:
Feat: 결제 기능 비즈니스 로직 작성

zmfpdl64 committed:
Feat: 결제 기능 Repository 작성

zmfpdl64 committed:
Refactor: 매개변수 변경 리펙토링

zmfpdl64 committed:
Refactor: Dto 메시지 분리

Joowon-Seo commented on payment-api/src/main/java/org/ecommerce/paymentapi/internal/service/PaymentService.java:
```diff
@@ -0,0 +1,91 @@
+	 *
+	 * @param - PaymentPrice 주문 결제 요청 객체
+	 * @return - 반환 값 설명 텍스트
+	 */
+	@DistributedLock(key = {
+		"'BEANPAY'.concat(#paymentRollBack.sellerId()).concat('SELLER')",
+		"'BEANPAY'.concat(#paymentRollBack.userId()).concat('USER')",
+	})
+	public PaymentDto paymentPriceRollBack(final PaymentRollBack paymentRollBack) {
+		Payment payment = getPayment(paymentRollBack);
+		return PaymentMapper.INSTANCE.toDto(payment.rollbackPayment());
+	}
+
+	private Payment getPayment(PaymentRollBack paymentRollBack) {
```
결제와 환불에서 OrderId를 사용하고 계신 것 같습니다.
Order Table의 Id를 의미하는 것 인가요?,
환불에서 사용하는 OrderId가 OrderDetailId를 의미한다면 '부분 환불'임으로 큰 문제는 없을 것 같습니다. 👍 

저희 비즈니즈는 Order안에 OrderDetails가 존재하고,
결제 부분에서 Order에 대한 OrderDetails의 결제를 하나의 트랜잭션으로 묶어서 진행하는 '일괄결제'로 알고 있습니다.
`paymentPrice`에서는 단일 OrderDetail의 결제만 이루어지는 것 같습니다.
만약 `paymentPrice`에서 사용하는 OrderId가 OrderDetailId를 의미하는 것이라면 이를 호출하는 상위 메소드가 필요할 것 같습니다. 😄

Joowon-Seo replied:
`"'BEANPAY'.concat(#paymentRollBack.sellerId()).concat('SELLER')"` 상수로 관리하시면 더욱 안정적이고 가독성이 좋을 것 같습니다! 👍

zmfpdl64 replied:
맞습니다 OrderDetail마다 다른 Seller를 가지고 있는데 현재는 한 셀러와만 계산하네요 감사합니다 👍

zmfpdl64 replied:
상수로 관리하는게 훨씬 가독성이 좋을것 같습니다 감사합니다 주원님!
락로직이 고정되면 변경하도록 하겠습니다!

Joowon-Seo commented:
멀티락 구현이 인상 깊었습니다. 👍

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/dto/PaymentDto.java:
```diff
@@ -1,8 +1,53 @@
+		) {
+		}
+
+		public record PaymentRollBack(
+			@Min(value = 1, message = "1이상의 주문ID를 전달해주세요")
+			Long orderId,
+			@Min(value = 1, message = "1이상의 유저ID를 전달해주세요")
+			Integer userId,
+			@Min(value = 1, message = "1이상의 셀러ID를 전달해주세요")
+			Integer sellerId
+		) {
```
PaymentRollBack 의 에러메시지를 실수로 문자열로 직접 넣으신 것 같습니다!

zmfpdl64 replied:
감사합니다 원호님 수정해서 올리겠습니다!

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/repository/impl/BeanPayRepositoryImpl.java:
```diff
@@ -24,14 +24,20 @@ public class BeanPayRepositoryImpl implements BeanPayCustomRepository {
 
 		return Optional.ofNullable(jpaQueryFactory.selectFrom(beanPay)
-			.where(beanPay.userId.eq(userId).and(beanPay.role.eq(role)))
+			.where(
+				beanPay.userId.eq(userId)
+				.and(
+				beanPay.role.eq(role)))
 			.fetchOne());
```
이거는 도영님께도 여쭤봐야하는데 queryDsl 을 사용할때 Where절을 BooleanExpression 메서드로 만드는게 null 이면 무시하면서 가독성도 더 좋을 것 같습니다 오늘 멘토링 시간에 여쭤보도록 하겠습니다

zmfpdl64 replied:
BooleanExpression을 사용하는 방식도 있군요 좋은 정보 감사합니다!

Joowon-Seo replied:
원호이 말씀주셨던 것 처럼 Default가 and 조건이라 `and` -> `,`로 변경하시면 될 것 같습니다.

no-oneho reviewed (approved):
고생많으셨습니다 우진님!!

hjm12223 reviewed (approved):
수고하셨습니다 우진님!

zmfpdl64 committed:
Refactor: Dto Message 분리

zmfpdl64 committed:
Refactor: 결제하기 롤백로직

zmfpdl64 committed:
Refactor: 결제 리펙토링

zmfpdl64 committed:
Refactor: Integer Long으로 수정

zmfpdl64 committed:
Refactor: 에러코드 수정

결제 디테일의 상태가 현재 진행상태와 결제, 환불 상태를 나타내는 Enum으로 분리

빈페이의 값이 부족함이BeanPayErrorCode가 더 적절해 수정함

zmfpdl64 committed:
Refactor: id로 맵핑방식 Map방식 List로 변경

기존 방식은 중복된 sellerid가 존재하면 기존의 key를 삭제하여 결제가 원하는대로 진행되지 않음

zmfpdl64 committed:
Refactor: Mapper 필드 맵핑

zmfpdl64 committed:
chore: mock 데이터 추가

zmfpdl64 committed:
Test: 결제 진행 롤백 로직 작성

zmfpdl64 committed:
Merge branch 'develop' of https://github.com/Team-Koreano/Koreano into feature/payment/#86

zmfpdl64 committed:
Refactor: Dto 이름 변경

zmfpdl64 committed:
Refactor: Test Controller 분리

zmfpdl64 committed:
Feat: BeanPayDto 작성

zmfpdl64 committed:
Feat: 빈페이 생성

zmfpdl64 committed:
Test: 빈페이 생성 테스트

zmfpdl64 committed:
Chore: 주석추가

zmfpdl64 committed:
Feat: 멀티 가변락 작성

zmfpdl64 committed:
Refactor: 락 이름 상수로 분리

zmfpdl64 committed:
Refactor: 연관 Entity 종속성 추가

zmfpdl64 committed:
Refactor: 테스트에 필요한 데이터 추가

zmfpdl64 committed:
Refactor: MapStruct Setter -> AllArgsStructor 변경

zmfpdl64 committed:
Refactor: +연산 concat으로 변경

userId는 Integer여서 +연산으로 변경
그 외의 문자열을 붙히는 것은 concat으로 수행

zmfpdl64 committed:
Merge pull request #150 from Team-Koreano/feature/payment/#148

[결제] Spring AOP List 가변 락 적용

zmfpdl64 committed:
Merge branch 'develop' of https://github.com/Team-Koreano/Koreano into feature/payment/#86

zmfpdl64 committed:
Merge branch 'refs/heads/feature/payment/#86' into feature/payment/#146

zmfpdl64 committed:
Merge branch 'feature/payment/#146' of https://github.com/Team-Koreano/Koreano into feature/payment/#146

zmfpdl64 committed:
Merge pull request #147 from Team-Koreano/feature/payment/#146

[결제] 빈페이 생성 구현
